File: src/test/java/com/gemini/krakenbot/service/PortfolioManagerOrderExecutionTest.java
package com.gemini.krakenbot.service;

import java.math.BigDecimal;

import com.gemini.krakenbot.model.PortfolioStats;
import com.gemini.krakenbot.service.impl.*;

import com.gemini.krakenbot.config.Allocation;
import com.gemini.krakenbot.config.AppConfig;
import com.gemini.krakenbot.config.Settings;
import com.gemini.krakenbot.repository.PortfolioStatsRepository;
import com.gemini.krakenbot.repository.impl.PortfolioStatsRepositoryImpl;
import org.junit.jupiter.api.Test;
import org.mockito.InOrder;
import org.springframework.test.util.ReflectionTestUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.mockito.Mockito.*;

class PortfolioManagerOrderExecutionTest {

        @Test
        void testExecutionOrder_SellsBeforeBuys() {
                // Setup
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));
                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService,
                                repo);

                // Config Mock
                // Asset A: Overweight -> Needs Sell
                // Asset B: Underweight -> Needs Buy
                Allocation allocA = new Allocation("A", 10.0);
                Allocation allocB = new Allocation("B", 90.0);
                Allocation allocUSD = new Allocation("USD", 0.0); // No cash target

                List<Allocation> allAllocations = List.of(allocA, allocB, allocUSD);
                AppConfig mockConfig = mock(AppConfig.class);
                Settings mockSettings = mock(Settings.class);

                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(mockConfig.settings()).thenReturn(mockSettings);
                when(mockSettings.deviationTriggerPercent()).thenReturn(1.0); // 1% trigger
                when(mockSettings.dustThresholdUSD()).thenReturn(1.0);

                when(configService.getConfig()).thenReturn(mockConfig);

                // Balances Mock
                // Total Value = $1000
                // A: has $500 (50%) -> Target 10% ($100) -> Sell $400
                // B: has $500 (50%) -> Target 90% ($900) -> Buy $400
                // USD: $0
                Map<String, Double> balances = new HashMap<>();
                balances.put("A", 5.0); // Price 100 -> $500
                balances.put("B", 50.0); // Price 10 -> $500
                balances.put("USD", 0.0);
                when(krakenService.getBalances()).thenReturn(balances);

                // Prices Mock
                Map<String, Double> prices = new HashMap<>();
                prices.put("AUSD", 100.0);
                prices.put("BUSD", 10.0);
                when(krakenService.getTickerPrices(anyString())).thenReturn(prices);

                // Invoke private method
                ReflectionTestUtils.invokeMethod(portfolioManager, "performRebalanceCycle");

                // Verify Order
                InOrder inOrder = inOrder(krakenService);

                // 1. Should call Sell for A
                inOrder.verify(krakenService).executeOrder(eq("AUSD"), eq("market"), eq("sell"), anyDouble());

                // 2. Should call Buy for B
                // Note: It will only buy B if it thinks it has cash.
                // The code adds sold value to projectedCash immediately.
                // A Sell $400 -> Projected Cash $400 -> Buy B $400.
                inOrder.verify(krakenService).executeOrder(eq("BUSD"), eq("market"), eq("buy"), anyDouble());
        }

        @Test
        void testExecution_SkipDustSells() {
                // Setup
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));

                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService,
                                repo);

                // Config Mock
                // Asset A: Overweight -> Needs Sell
                Allocation allocA = new Allocation("A", 10.0);
                Allocation allocUSD = new Allocation("USD", 90.0);

                List<Allocation> allAllocations = List.of(allocA, allocUSD);
                AppConfig mockConfig = mock(AppConfig.class);
                Settings mockSettings = mock(Settings.class);

                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(mockConfig.settings()).thenReturn(mockSettings);
                when(mockSettings.deviationTriggerPercent()).thenReturn(0.1);
                when(mockSettings.dustThresholdUSD()).thenReturn(10.0); // Threshold 10

                when(configService.getConfig()).thenReturn(mockConfig);

                // Balances Mock
                // Total Value = $1000
                // A: has $105 (10.5%) -> Target 10% ($100) -> Sell $5. (5 < 10) -> Skip
                // USD: $895
                Map<String, Double> balances = new HashMap<>();
                balances.put("A", 1.05); // Price 100 -> $105
                balances.put("USD", 895.0);
                when(krakenService.getBalances()).thenReturn(balances);

                // Prices Mock
                Map<String, Double> prices = new HashMap<>();
                prices.put("AUSD", 100.0);
                when(krakenService.getTickerPrices(anyString())).thenReturn(prices);

                // Invoke private method
                ReflectionTestUtils.invokeMethod(portfolioManager, "performRebalanceCycle");

                // Verify NO Sell
                verify(krakenService, never()).executeOrder(eq("AUSD"), anyString(), eq("sell"), anyDouble());
        }

        @Test
        void testExecution_CashVerificationFallback() {
                // Setup
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));
                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService, repo);

                List<Allocation> allAllocations = List.of(new Allocation("A", 10.0), new Allocation("B", 90.0));
                AppConfig mockConfig = mock(AppConfig.class);
                Settings mockSettings = mock(Settings.class);
                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(mockConfig.settings()).thenReturn(mockSettings);
                when(mockSettings.deviationTriggerPercent()).thenReturn(1.0);
                when(mockSettings.dustThresholdUSD()).thenReturn(1.0);
                when(configService.getConfig()).thenReturn(mockConfig);

                // Initial balances
                Map<String, Double> initialBalances = new HashMap<>();
                initialBalances.put("A", 5.0); // $500
                initialBalances.put("B", 50.0); // $500
                initialBalances.put("USD", 0.0);

                // First call: returns initial balances
                // Second call (verification): throws Exception
                when(krakenService.getBalances())
                    .thenReturn(initialBalances)
                    .thenThrow(new RuntimeException("API Error during verification!"));

                Map<String, Double> prices = new HashMap<>();
                prices.put("AUSD", 100.0);
                prices.put("BUSD", 10.0);
                when(krakenService.getTickerPrices(anyString())).thenReturn(prices);

                // Execute
                ReflectionTestUtils.invokeMethod(portfolioManager, "performRebalanceCycle");

                // It should fall back to projectedCash ($400 from sell A) and buy B
                verify(krakenService).executeOrder(eq("AUSD"), eq("market"), eq("sell"), anyDouble());
                verify(krakenService).executeOrder(eq("BUSD"), eq("market"), eq("buy"), anyDouble());
        }

        @Test
        void testExecution_PartialFillCashUpdate() {
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));
                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService, repo);

                List<Allocation> allAllocations = List.of(new Allocation("A", 10.0), new Allocation("B", 90.0));
                AppConfig mockConfig = mock(AppConfig.class);
                Settings mockSettings = mock(Settings.class);
                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(mockConfig.settings()).thenReturn(mockSettings);
                when(mockSettings.deviationTriggerPercent()).thenReturn(1.0);
                when(mockSettings.dustThresholdUSD()).thenReturn(1.0);
                when(configService.getConfig()).thenReturn(mockConfig);

                Map<String, Double> initialBalances = new HashMap<>();
                initialBalances.put("A", 5.0); // $500 -> sell $400
                initialBalances.put("B", 50.0); // $500 -> buy $400
                initialBalances.put("USD", 0.0);

                Map<String, Double> updatedBalances = new HashMap<>();
                updatedBalances.put("A", 2.0); // sold some
                updatedBalances.put("B", 50.0); 
                updatedBalances.put("USD", 200.0); // Only got $200! (Partial fill/slippage)

                when(krakenService.getBalances())
                    .thenReturn(initialBalances)
                    .thenReturn(updatedBalances);

                Map<String, Double> prices = new HashMap<>();
                prices.put("AUSD", 100.0);
                prices.put("BUSD", 10.0);
                when(krakenService.getTickerPrices(anyString())).thenReturn(prices);

                ReflectionTestUtils.invokeMethod(portfolioManager, "performRebalanceCycle");

                verify(krakenService).executeOrder(eq("AUSD"), eq("market"), eq("sell"), anyDouble());
                // It should buy B but only with $200 cash * 0.99 = $198. $198 / 10 = 19.8 units
                verify(krakenService).executeOrder(eq("BUSD"), eq("market"), eq("buy"), doubleThat(v -> Math.abs(v - 19.8) < 0.1));
        }
}

File: src/test/kotlin/com/gemini/krakenbot/service/PortfolioManagerOrderExecutionTest.kt
package com.gemini.krakenbot.service

import com.gemini.krakenbot.TestFixtures
import com.gemini.krakenbot.config.Allocation
import com.gemini.krakenbot.model.Asset
import com.gemini.krakenbot.model.PortfolioStats
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.comparables.shouldBeEqualComparingTo
import io.kotest.matchers.shouldBe
import io.mockk.coEvery
import io.mockk.every
import kotlinx.coroutines.test.runTest
import java.math.BigDecimal

class PortfolioManagerOrderExecutionTest : StringSpec() {

    override fun isolationMode() = IsolationMode.InstancePerTest

    private lateinit var fixture: PortfolioManagerTestFixture

    init {
        beforeTest {
            fixture = createPortfolioManagerTestFixture()
            coEvery {
                fixture.portfolioStatsRepository.load()
            } returns PortfolioStats(BigDecimal.ZERO)
        }

        "testExecutionOrder_SellsBeforeBuys" {
            runTest {
                val allocA = Allocation("A", 10.0)
                val allocB = Allocation("B", 90.0)
                val allocUSD = Allocation(
                    Asset.USD,
                    0.0,
                )
                val allAllocations = listOf(allocA, allocB, allocUSD)

                val mockSettings = TestFixtures.settings(dryRun = false, deviationTriggerPercent = 1.0)
                val mockConfig = TestFixtures.config(
                    settings = mockSettings,
                    allocations = allAllocations,
                )

                every { fixture.configService.getConfig() } returns mockConfig

                fixture.krakenService.balanceSupplier = {
                    val sold = fixture.krakenService.executedOrders.any { it.side.equals("sell", ignoreCase = true) }
                    if (sold) {
                        mapOf("A" to 1.0, "B" to 50.0, Asset.USD to 400.0)
                    } else {
                        mapOf("A" to 5.0, "B" to 50.0, Asset.USD to 0.0)
                    }
                }

                val prices = mapOf("AUSD" to 100.0, "BUSD" to 10.0)
                fixture.krakenService.pricesSupplier = { prices }

                fixture.portfolioManager.performRebalanceCycle()

                fixture.krakenService.executedOrders.size shouldBe 2
                fixture.krakenService.executedOrders[0].pair shouldBe "AUSD"
                fixture.krakenService.executedOrders[0].side shouldBe "sell"
                fixture.krakenService.executedOrders[1].pair shouldBe "BUSD"
                fixture.krakenService.executedOrders[1].side shouldBe "buy"
            }
        }

        "testExecution_SkipDustSells" {
            runTest {
                val allocA = Allocation("A", 10.0)
                val allocUSD = Allocation(
                    Asset.USD,
                    90.0,
                )
                val allAllocations = listOf(allocA, allocUSD)

                val mockSettings = TestFixtures.settings(
                    dryRun = false,
                    deviationTriggerPercent = 0.1,
                    dustThresholdUSD = 10.0,
                )
                val mockConfig = TestFixtures.config(
                    settings = mockSettings,
                    allocations = allAllocations,
                )

                every { fixture.configService.getConfig() } returns mockConfig

                val balances = mapOf("A" to 1.05, Asset.USD to 895.0)
                fixture.krakenService.balanceSupplier = { balances }

                val prices = mapOf("AUSD" to 100.0)
                fixture.krakenService.pricesSupplier = { prices }

                fixture.portfolioManager.performRebalanceCycle()

                fixture.krakenService.executedOrders.none {
                    it.pair == "AUSD" && it.side == "sell"
                } shouldBe true
            }
        }

        "testExecution_CashVerificationFallback" {
            runTest {
                val allAllocations =
                    listOf(
                        Allocation(
                            "A",
                            10.0,
                        ),
                        Allocation(
                            "B",
                            90.0,
                        ),
                    )
                val mockSettings = TestFixtures.settings(dryRun = false, deviationTriggerPercent = 1.0)
                val mockConfig = TestFixtures.config(
                    settings = mockSettings,
                    allocations = allAllocations,
                )
                every { fixture.configService.getConfig() } returns mockConfig

                val initialBalances =
                    mapOf("A" to 5.0, "B" to 50.0, Asset.USD to 0.0)

                var callCount = 0
                fixture.krakenService.balanceSupplier = {
                    callCount++
                    if (callCount == 1) {
                        initialBalances
                    } else {
                        throw RuntimeException(
                            "API Error during verification!",
                        )
                    }
                }

                val prices = mapOf("AUSD" to 100.0, "BUSD" to 10.0)
                fixture.krakenService.pricesSupplier = { prices }

                fixture.portfolioManager.performRebalanceCycle()

                // Fail-closed: balance polls throw → no positive USD observed → buys aborted
                fixture.krakenService.executedOrders.size shouldBe 1
                fixture.krakenService.executedOrders[0].pair shouldBe "AUSD"
                fixture.krakenService.executedOrders[0].side shouldBe "sell"
            }
        }

        "testExecution_PartialFillCashUpdate" {
            runTest {
                val allAllocations =
                    listOf(
                        Allocation(
                            "A",
                            10.0,
                        ),
                        Allocation(
                            "B",
                            90.0,
                        ),
                    )
                val mockSettings = TestFixtures.settings(dryRun = false, deviationTriggerPercent = 1.0)
                val mockConfig = TestFixtures.config(
                    settings = mockSettings,
                    allocations = allAllocations,
                )
                every { fixture.configService.getConfig() } returns mockConfig

                val initialBalances =
                    mapOf("A" to 5.0, "B" to 50.0, Asset.USD to 0.0)
                val updatedBalances =
                    mapOf("A" to 2.0, "B" to 50.0, Asset.USD to 200.0)

                var callCount = 0
                fixture.krakenService.balanceSupplier = {
                    callCount++
                    if (callCount == 1) initialBalances else updatedBalances
                }

                val prices = mapOf("AUSD" to 100.0, "BUSD" to 10.0)
                fixture.krakenService.pricesSupplier = { prices }

                fixture.portfolioManager.performRebalanceCycle()

                fixture.krakenService.executedOrders.size shouldBe 2
                fixture.krakenService.executedOrders[0].pair shouldBe "AUSD"
                fixture.krakenService.executedOrders[0].side shouldBe "sell"
                fixture.krakenService.executedOrders[1].pair shouldBe "BUSD"
                fixture.krakenService.executedOrders[1].side shouldBe "buy"
                (
                    fixture.krakenService.executedOrders[1].volume.subtract(
                        BigDecimal.valueOf(
                            19.8,
                        ),
                    ).abs() < BigDecimal("0.1")
                    ).shouldBeTrue()
            }
        }

        "testExecution_BuyVolumeFloorsAtCryptoPrecision" {
            runTest {
                val actionLog = mutableListOf<String>()

                fixture.orderExecutor.executeOrders(
                    buyOrders = mapOf(Asset.BTC to BigDecimal("1.00")),
                    sellOrders = emptyMap(),
                    currentValuesUSD = mapOf(Asset.USD to BigDecimal("10.00")),
                    prices = mapOf(Asset.BTC to BigDecimal("6.00")),
                    settings = TestFixtures.settings(
                        dryRun = false,
                        dustThresholdUSD = 0.99,
                        simulation = true,
                    ),
                    actionLog = actionLog,
                    cycleId = "buy-volume-floor",
                    availableBalances = null,
                )

                val order = fixture.krakenService.executedOrders.single()
                order.volume.shouldBeEqualComparingTo(BigDecimal("0.16666666"))
                (order.volume.multiply(BigDecimal("6.00")) <= BigDecimal("1.00")).shouldBeTrue()
            }
        }
    }
}
